s3/src/main/scala/geotrellis/spark/io/s3/S3RDDWriter.scala

```diff
@@ -56,42 +77,80 @@ trait S3RDDWriter {
-              val is = new ByteArrayInputStream(bytes)
-              val request = putObjectModifier(new PutObjectRequest(bucket, key, is, metadata))
-              Some(request, iter)
-            } else {
-              None
-            }
-          }
+              val record = iter.next()
+              val key = record._1
+              val kvs = record._2.toVector
+              Some(((key, kvs), iter))
+            } else None
+          })
+
+        def elaborateRow(row: (String, Vector[(K,V)])): Process[Task, (String, Vector[(K,V)])] = {
+          Process eval Task({
+            val (key, kvs1) = row
+            val kvs2: Vector[(K,V)] =
+              if (mergeFunc != None) {
+                try {
+                  val bytes = IOUtils.toByteArray(s3client.getObject(bucket, key).getObjectContent)
+                  AvroEncoder.fromBinary(schema, bytes)(_recordCodec)
+                } catch {
+                  case e: AmazonS3Exception if e.getStatusCode == 404 => Vector.empty
+                }
+              } else Vector.empty
+            val kvs =
+              mergeFunc match {
+                case Some(fn) =>
+                  (kvs2 ++ kvs1)
+                    .groupBy({ case (k,v) => k })
+                    .map({ case (k, kvs) =>
+                      val vs = kvs.map({ case (k,v) => v }).toSeq
+                      val v: V = vs.tail.foldLeft(vs.head)(fn)
+                      (k, v) })
+                    .toVector
+                case None => kvs1
+              }
+            (key, kvs)
+          })(pool)
+        }
 
-        val pool = Executors.newFixedThreadPool(threads)
+        def rowToRequest(row: (String, Vector[(K,V)])): Process[Task, PutObjectRequest] = {
+          Process eval Task({
+            val (key, kvs) = row
+            val bytes = AvroEncoder.toBinary(kvs)(_codec)
+            val metadata = new ObjectMetadata()
+            metadata.setContentLength(bytes.length)
+            val is = new ByteArrayInputStream(bytes)
+            putObjectModifier(new PutObjectRequest(bucket, key, is, metadata))
+          })(pool)
+        }
 
-        val write: PutObjectRequest => Process[Task, PutObjectResult] = { request =>
-          Process eval Task {
+        def retire(request: PutObjectRequest): Process[Task, PutObjectResult] = {
+          Process eval Task({
             request.getInputStream.reset() // reset in case of retransmission to avoid 400 error
             s3client.putObject(request)
-          }(pool).retryEBO {
+          })(pool).retryEBO {
             case e: AmazonS3Exception if e.getStatusCode == 503 => true
             case _ => false
           }
         }
 
-        val results = nondeterminism.njoin(maxOpen = threads, maxQueued = threads) { requests map write }(Strategy.Executor(pool))
+        val results = nondeterminism.njoin(maxOpen = threads, maxQueued = threads) {
+          rows flatMap elaborateRow flatMap rowToRequest map retire
```

Comment:
Can't we abstract over it a bit more? Cassandra uses the same thing; or it's hardly possible?